src/gt4py/stencil_builder.py

```diff
@@ -60,14 +61,17 @@ def __init__(
             backend(self) if backend else gt4py.backend.from_name("debug")(self)
         )
         self.frontend: "FrontendType" = frontend or gt4py.frontend.from_name("gtscript")
-        self.caching = gt4py.caching.strategy_factory("jit", self)
+        self.caching = self.get_caching_strategy(options)
         self._build_data: Dict[str, Any] = {}
         self._externals: Dict[str, Any] = {}
 
     def build(self) -> Type["StencilObject"]:
         """Generate, compile and/or load everything necessary to provide a usable stencil class."""
-        # load or generate
-        stencil_class = None if self.options.rebuild else self.backend.load()
+        # load, defer, or generate
+        if self.caching.is_deferred():
+            stencil_class = self.caching.defer()
+        else:
+            stencil_class = None if self.options.rebuild else self.backend.load()
```

Comment:
What stops us from simply checking here if an option "deferred_building_callback" was passed to the builder and if so, then returning the result of calling the callback with `self` instead of the stencil class?

Reply:
That would simplify the design considerably